torch/csrc/api/include/torch/nn/module.h

```diff
@@ -87,13 +87,70 @@ class Module {
   }
 
  protected:
-  std::shared_ptr<nn::Module> add(
-      std::shared_ptr<nn::Module>,
-      std::string const&);
-  // Be careful when registering Tensors that are not variables
-  Variable& add(Variable, std::string const&);
+  using ModulePtr = std::shared_ptr<nn::Module>;
+
+  template <typename Derived, typename M>
+  void register_module(
+      const std::string& name,
+      std::shared_ptr<M> Derived::*module,
+      std::shared_ptr<M> new_module) {
+    check_name("Module", name);
+    auto base_module =
+        reinterpret_cast<std::shared_ptr<nn::Module> Derived::*>(module);
+    const auto pair = children_.emplace(name, base_module);
+    AT_CHECK(pair.second, "Module " + name + " has already been registered");
+    (pair.first->second)(static_cast<Derived*>(this)) = std::move(new_module);
+  }
+
+  // Hack for Sequential, should remove/figure out to do this better
+  template <typename Derived>
+  void register_module(
+      const std::string& name,
+      std::vector<std::shared_ptr<Module>> Derived::*modules,
+      size_t index) {
+    const auto pair = children_.insert({name, {modules, index}});
+    AT_CHECK(pair.second, "Module " + name + " has already been registered");
+  }
+
+  template <typename Derived>
+  void register_parameter(
+      const std::string& name,
+      Variable Derived::*variable,
+      Tensor tensor) {
+    check_name("Parameter", name);
+    const auto pair = parameters_.emplace(name, variable);
+    AT_CHECK(pair.second, "Parameter " + name + " has already been registered");
+    (pair.first->second)(static_cast<Derived*>(this)) =
+        autograd::make_variable(tensor, /*requires_grad=*/true);
+  }
+
+  template <typename Derived>
+  void register_parameters(ParameterList Derived::*parameter_list) {
```

Comment:
Meh, can we implement this as an RNN method? When would that be useful to people in general?

Reply:
There's `ParameterList` in Python PyTorch for this, so wouldn't it be useful generally?

Reply:
Well but `ParameterList` is meant to be used as a submodule. There's no such thing as "copy all parameters from that module to this module". That is inconsistent with Python, and I don't think it fits the API that well.